File: concurrency/concurrent_mutex.go
package main

import (
	"fmt"
	"sync"
)

func main() {
	var mutex sync.Mutex
	var counter int = 0
	var wg sync.WaitGroup

	orderedTask := func(taskID int) {
		defer wg.Done()

		mutex.Lock() // Acquire mutex to ensure exclusive access
		defer mutex.Unlock()

		counter++
		fmt.Printf("Task %d executing as number %d\n", taskID, counter)
		// Database operation would go here
		// db.Exec("UPDATE table SET value = ? WHERE id = ?", counter, someID)
	}

	// Launch goroutines
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go orderedTask(i)
	}

	wg.Wait()
}

File: concurrency/concurrent_mutex.py
import threading

# Using a lock to ensure ordered execution
lock = threading.Lock()
counter = 0


def ordered_task(task_id):
    global counter
    with lock:  # Acquire lock, ensuring only one thread executes this block at a time
        current = counter
        counter += 1
        print(f"Task {task_id} executing as number {counter}")
        # Database operation would go here
        # db.execute("UPDATE table SET value = %s WHERE id = %s", (counter, some_id))


# Launch threads
threads = []
for i in range(10):
    t = threading.Thread(target=ordered_task, args=(i,))
    threads.append(t)
    t.start()

for t in threads:
    t.join()
